controllers/mdutil/util.go

```diff
@@ -425,14 +427,57 @@ func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*cluste
 			return msList[i]
 		}
 	}
-	// new MachineSet does not exist.
+	// New MachineSet does not exist.
 	return nil
 }
 
+// FindNewMachineSetWithRolloutAfter is as FindNewMachineSet but it also accounts for RolloutAfter.
+func FindNewMachineSetWithRolloutAfter(now *metav1.Time, deployment *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (*clusterv1.MachineSet, error) {
+	// In rare cases, such as after cluster upgrades, Deployment may end up with
+	// having more than one new MachineSets that have the same template,
+	// see https://github.com/kubernetes/kubernetes/issues/40415
+	// We deterministically choose the oldest new MachineSet with matching template hash.
+	sort.Sort(MachineSetsByCreationTimestamp(msList))
+
+	if deployment.Spec.RolloutAfter != nil {
+		if now.After(deployment.Spec.RolloutAfter.Time) {
+			// TODO (alberto): inlining the format statement within the annotation assigment here
+			// does not set the value. Why?
+			LastRolloutAfter := deployment.Spec.RolloutAfter.Time.UTC().Format(time.RFC3339)
+			deployment.Annotations[clusterv1.LastRolloutAfterAnnotation] = LastRolloutAfter
```

Comment:
Nit: Don't think this needs to be exported
```suggestion
			lastRolloutAfter := deployment.Spec.RolloutAfter.Time.UTC().Format(time.RFC3339)
			deployment.Annotations[clusterv1.LastRolloutAfterAnnotation] = lastRolloutAfter
```